posts/scala-for-the-arrogant-part-2.md

```diff
@@ -0,0 +1,427 @@
+                .collect(Collectors.toSet())) : 
+        //Found a solution or no solution available
+        trails.stream().filter(SolutionExistsInTrail).findAny();
+}
+
+/**
+    * For the given parameters, find a solution where the Jug capacities can produce
+    * the goal amount.
+    * @return A list of actions that yield the goal amount, or null if no solution exists
+    */
+public Optional<List<Action>> findSolution() {
+    return findSolution(Collections.singletonList(startActionTrail), Collections.singleton(startState));
+}
+
+```
+We use a stream transformation [Function](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html) to take an Action Trail (defined as a list of actions) and turn it into the list of possible next actions. We then flatMap into a single stream for the same reasons described above. We use some similar architecture to modify our history. Rather than adding elements to the set we take our newly found trails, transform them into the end state of the last element, and add that state to the history if it does not yet exist.
+
+Finally, the `solutionExistsInTrail` filter defines a [Predicate](https://docs.oracle.com/javase/8/docs/api/java/util/function/Predicate.html) that gives a concise and clear way to identify a solution.
+
+## Evaluating Our Design
+[Review the commit](https://github.com/randeepbydesign/diehardJugSolver/commit/63d79218996856317a371700b8b757c89d4f1fd4) for a Java **Functional** solution and you will see: 
+
+- There is no modifying collections, and therefore no modification of state. If we need to create a Set with an additional element, for example, we create a new set with the original elements and the new element to be added. 
+
+- There are no for/while loops in this solution. Rather we have replaced them with Functional logic to filter, transform, and collect our streams of data. The final form of this solution is much easier to write Unit tests for as well as we have decomposed the logic into functions that take concise and discrete inputs and return specific consistent values.
+
+- Our semicolon key gets a rest as the majority of our functions are one-liners. Our method names are descriptive enough that composing them together adds readability to our solution. Admittedly, it can be more difficult to debug, but I would argue that structuring your programs in this manner creates a final result that contains far fewer defects than traditional methods.
+
+One critique you may have is that there is still quite a bit of verbosity to the code. For example, having a method that returns a Predicate can look unwieldy- especially if you are not familiar with that type of syntax. We also manage a parameterized List of Lists of Actions that we define as an Action Trail. We also manage our history as a Set of ArrayLists of Integers (we opt for ArrayList over generic List just to enforce the idea that the sequence of elements is important).
```

Comment:
"we opt for ArrayList over generic List just to enforce the idea that the sequence of elements is important" -> By definition, a List is a sequence. This is the first line of documentation in https://docs.oracle.com/javase/8/docs/api/java/util/List.html:

> An ordered collection (also known as a sequence).

Reply:
resolved